apis/ninedof/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+            data_cell.set(Some(data));
+        });
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::read_accelerometer() {
+                    while data_cell.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match data_cell.get() {
+            None => Err(ErrorCode::Busy),
+            Some(data) => {
+                *x = data.x;
+                *y = data.y;
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Initiate a synchronous magnetometer measurement.
+    /// Returns Ok(data) if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_magnetometer_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
+        let data_cell: Cell<Option<NineDofData>> = Cell::new(None);
+        let listener = NineDofListener(|data| {
+            data_cell.set(Some(data));
+        });
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::read_magnetometer() {
+                    while data_cell.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match data_cell.get() {
+            None => Err(ErrorCode::Busy),
+            Some(data) => {
+                *x = data.x;
+                *y = data.y;
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Initiate a synchronous gyroscope measurement.
+    /// Returns Ok(data) as NineDofData if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_gyro_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
+        let data_cell: Cell<Option<NineDofData>> = Cell::new(None);
+        let listener = NineDofListener(|data| {
+            data_cell.set(Some(data));
+        });
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::read_gyro() {
+                    while data_cell.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match data_cell.get() {
+            None => Err(ErrorCode::Busy),
+            Some(data) => {
+                *x = data.x;
+                *y = data.y;
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Calculate the magnitude of the accelerometer reading
+    /// Returns value of magnitude if the operation was successful
+    /// Returns 0.0 if the operation was unsuccessful
+    pub fn ninedof_read_accel_mag() -> f64 {
```

Comment:
Magnitudes are valid only for the accelerometer?

Reply:
I haven't seen them implemented in libtock-c for the [9-DOF](https://github.com/tock/libtock-c/blob/master/libtock/ninedof.c#L24), I assume that for the magnetometer and gyroscope the magnitudes do not hold any information that could be useful, but for the accelerometer you could use it to get the total acceleration

Reply:
Ok.